watcher/ops.go

```diff
@@ -66,18 +69,48 @@ func listDirRec(dir string) ([]string, error) {
 	return files, nil
 }
 
-func debounce(delay time.Duration, input chan fsnotify.Event) chan *fsnotify.Event {
+func dedup(delay time.Duration, input chan fsnotify.Event, dedup bool) chan *fsnotify.Event {
 	output := make(chan *fsnotify.Event)
+	var mu sync.Mutex
+	timers := make(map[string]*time.Timer)
+
 	go func() {
 		for {
 			select {
 			case event, ok := <-input:
 				if !ok {
 					return
 				}
-				output <- &event
-			case <-time.After(50 * time.Millisecond):
-				time.Sleep(delay)
+
+				if !dedup {
+					output <- &event
+					continue
+				}
+
+				filepath := strings.TrimPrefix(event.Name, "./")
+
+				mu.Lock()
+				t, ok := timers[filepath]
+				mu.Unlock()
+
+				// If there's no timer, create one
+				if !ok {
+					t = time.AfterFunc(math.MaxInt64, func() {
```

Comment:
@borkdude 
>  Does this also mean that you will never see a duplicate event even after a long time?

This callback gets triggered after a certain interval. It puts the fs notification in the output channel and then deletes the timer. After this interval the notifications for that file resume.

Reply:
I'd like to know how the waiting for a very long time is helping here, not sure i can follow 😅 hopefully the example for fsnotify should help or some more explanations?

Reply:
I think it's just to initiate the timer "at infinite":

https://github.com/fsnotify/fsnotify/blob/main/cmd/fsnotify/dedup.go#L91

Immediately after, the timer is stopped and reset so unless I'm missing something, it could be set at `delay`?

Reply:
right, so this looks like an elaborate way to simply acquire the lock to safely add the timer to the map: create an infinite timer as we are not sure if the map could be changed within the delay duration. given Go's semantics, this seems okay.

Reply:
https://pkg.go.dev/sync?utm_source=godoc#Map could be a slightly less ceremony alternative but not sure if its worth it.